### PR TITLE
add shift on BartForCausalLM

### LINKED ISSUE #10480: Different result in AutoModelForCausalLM
# 🚀 Feature request
Models inside AutoModelForCausalLM have different behavior on loss calculation.

In BartForCausalLM
there is no shift in loss calculation
https://github.com/huggingface/transformers/blob/b013842244df7be96b8cc841491bd1e35e475e36/src/transformers/models/bart/modeling_bart.py#L1745
```
loss_fct = CrossEntropyLoss()
loss = loss_fct(logits.view(-1, self.config.vocab_size), labels.view(-1))
```

In RobertaForCausalLM
A shift is applied before loss calculation
https://github.com/huggingface/transformers/blob/b013842244df7be96b8cc841491bd1e35e475e36/src/transformers/models/roberta/modeling_roberta.py#L944
```
# we are doing next-token prediction; shift prediction scores and input ids by one
shifted_prediction_scores = prediction_scores[:, :-1, :].contiguous()
labels = labels[:, 1:].contiguous()
loss_fct = CrossEntropyLoss()
lm_loss = loss_fct(shifted_prediction_scores.view(-1, self.config.vocab_size), labels.view(-1))
```

## Motivation
I found a mistake when I switched the config from Roberta to BART in AutoModelForCausalLM. It turns out to be different labeling in loss. So, It would be nice to make CausalLM models handle label in the same way, either shift or not.
  

## Your contribution
I can make a PR to make sure that all the models will have a shift prediction.

### PR DESCRIPTION
**Pull request by voidful**
# What does this PR do?

Fixes  #10480

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

@patil-suraj 
@patrickvonplaten

### CODE REVIEW

File: src/transformers/models/bart/modeling_bart.py [hunk 1]
```diff
@@ -1723,6 +1723,9 @@ def forward(
         )
         return_dict = return_dict if return_dict is not None else self.config.use_return_dict
 
+        if labels is not None:
```
Comment (patrickvonplaten) [inline src/transformers/models/bart/modeling_bart.py:1726]: Hey @voidful,

Sorry this is not really what I meant. We should align the behavior of `BartForCausalLM` with `RobertaForCausalLM`. This change would mean that one can pass only `labels` and `input_ids` are created automatically - this is not how it's done in `GPT2LMHead`, `RobertaForCausalLM`, ... We're thinking a bit more on how to solve this internally with @patil-suraj - so we'll have to pause this PR for a bit

Comment (voidful) [reply src/transformers/models/bart/modeling_bart.py:1726]: It's OK. 
It may not be the best way and alleging the behavior of BartForCausalLM with RobertaForCausalLM should be a problem we need to deal with ?

I am confused with the design for token shift when calculating loss, input_ids and labels should be the same in xxxForCausalLM? 
or would it be better that moving label shift outside CausalLM and let the shifting determinate by users?
